x/twap/logic.go

```diff
@@ -243,11 +261,55 @@ func computeArithmeticTwap(startRecord types.TwapRecord, endRecord types.TwapRec
 		}
 		return endRecord.P1LastSpotPrice, err
 	}
+
+	if isArithmeticTwap {
+		return computeArithmeticTwap(startRecord, endRecord, quoteAsset), err
+	}
+	return computeGeometricTwap(startRecord, endRecord, quoteAsset), err
+}
+
+// computeArithmeticTwap computes and returns an arithmetic TWAP between
+// two records given the quote asset.
+func computeArithmeticTwap(startRecord types.TwapRecord, endRecord types.TwapRecord, quoteAsset string) sdk.Dec {
 	var accumDiff sdk.Dec
 	if quoteAsset == startRecord.Asset0Denom {
 		accumDiff = endRecord.P0ArithmeticTwapAccumulator.Sub(startRecord.P0ArithmeticTwapAccumulator)
 	} else {
 		accumDiff = endRecord.P1ArithmeticTwapAccumulator.Sub(startRecord.P1ArithmeticTwapAccumulator)
 	}
-	return types.AccumDiffDivDuration(accumDiff, timeDelta), err
+	timeDelta := endRecord.Time.Sub(startRecord.Time)
+	return types.AccumDiffDivDuration(accumDiff, timeDelta)
+}
+
+// computeGeometricTwap computes and returns a geometric TWAP between
+// two records given the quote asset.
+// The computation works as follows:
+// - compute arithmetic mean of logarithms of spot prices.
+// - exponentiate the result to get the geometric mean.
+// - if quoted asset is asset 1, take reciprocal of the exponentiated result.
+func computeGeometricTwap(startRecord types.TwapRecord, endRecord types.TwapRecord, quoteAsset string) sdk.Dec {
+	accumDiff := endRecord.GeometricTwapAccumulator.Sub(startRecord.GeometricTwapAccumulator)
+
+	timeDelta := endRecord.Time.Sub(startRecord.Time)
+	arithmeticMeanOfLogPrices := types.AccumDiffDivDuration(accumDiff, timeDelta)
+
+	geometrciMeanDenom0 := twapPow(arithmeticMeanOfLogPrices)
+	// N.B.: Geometric mean of recprocals is reciprocal of geometric mean.
+	// https://proofwiki.org/wiki/Geometric_Mean_of_Reciprocals_is_Reciprocal_of_Geometric_Mean
+	if quoteAsset == startRecord.Asset1Denom {
+		return sdk.OneDec().Quo(geometrciMeanDenom0)
+	}
+	return geometrciMeanDenom0
+}
+
+// twapLog returns the logarithm of the given spot price, base 2.
+// TODO: basic test
+func twapLog(price sdk.Dec) sdk.Dec {
+	return osmomath.BigDecFromSDKDec(price).LogBase2().SDKDec()
+}
+
+// twapPow exponentiates the geometricTwapMathBase to the given exponent.
+// TODO: basic test and benchmark.
+func twapPow(exponent sdk.Dec) sdk.Dec {
```

Comment:
just want to confirm we are exponentiating by e right? so that our result is something like `e^(exponent)`

Reply:
The base for both power and log is `2`. So it is $$2^{ArithmeticMean(log_2{Prices})}$$